File: backend/src/middleware/rateLimit.js
/**
 * Rate limiting middleware for Express
 */
function createRateLimiter(requestsPerMinute = 60) {
  const requests = new Map();

  return (req, res, next) => {
    // Skip rate limiting for health check
    if (req.path === '/health') return next();

    const clientIp = req.ip || req.socket.remoteAddress || 'unknown';
    const now = Date.now();
    const windowMs = 60 * 1000; // 1 minute

    // Get or create request list for this IP
    let reqTimes = requests.get(clientIp) || [];

    // Clean old requests
    reqTimes = reqTimes.filter(t => now - t < windowMs);

    // Check rate limit
    if (reqTimes.length >= requestsPerMinute) {
      return res.status(429).json({
        error: 'Rate limit exceeded',
        message: `Maximum ${requestsPerMinute} requests per minute allowed`,
      });
    }

    // Add current request
    reqTimes.push(now);
    requests.set(clientIp, reqTimes);

    next();
  };
}

module.exports = { createRateLimiter };

File: backend2/app/middleware/rate_limit.py
from fastapi import Request, HTTPException
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from collections import defaultdict
from datetime import datetime, timedelta
import asyncio


class RateLimitMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, requests_per_minute: int = 60):
        super().__init__(app)
        self.requests_per_minute = requests_per_minute
        self.requests: defaultdict[str, list[datetime]] = defaultdict(list)
        self.cleanup_task = None

    async def dispatch(self, request: Request, call_next):
        # Get client IP
        client_ip = request.client.host if request.client else "unknown"

        # Skip rate limiting for health check
        if request.url.path == "/health":
            return await call_next(request)

        now = datetime.utcnow()
        minute_ago = now - timedelta(minutes=1)

        # Clean old requests
        self.requests[client_ip] = [
            req_time for req_time in self.requests[client_ip]
            if req_time > minute_ago
        ]

        # Check rate limit
        if len(self.requests[client_ip]) >= self.requests_per_minute:
            return JSONResponse(
                status_code=429,
                content={
                    "error": "Rate limit exceeded",
                    "message": f"Maximum {self.requests_per_minute} requests per minute allowed"
                }
            )

        # Add current request
        self.requests[client_ip].append(now)

        response = await call_next(request)
        return response
